server/build_event_protocol/accumulator/accumulator.go

```diff
@@ -186,13 +182,11 @@ func (v *BEValues) AddEvent(event *build_event_stream.BuildEvent) error {
 			if u.Scheme == "bytestream" {
 				v.hasBytestreamTestActionOutputs = true
 
-				// To protect our backends from thrashing -- stop
-				// copying outputs if there are way too many. This can
-				// happen if a ruleset is buggy.
-				if len(v.testOutputURIs) >= maxPersistableArtifacts {
-					continue
+				if p.TestResult.GetStatus() == build_event_stream.TestStatus_PASSED {
+					v.passedTestOutputURIs = append(v.passedTestOutputURIs, u)
+				} else {
+					v.failedTestOutputURIs = append(v.failedTestOutputURIs, u)
```

Comment:
I think we want to keep enforcing the limit in the Accumulator, otherwise we  could have an unbounded number of bytestream URLs sitting in memory here and cause the server to OOM.

Something like:

```
if len(passed) + len(failed) >= limit {
  if len(passed) > 0 {
    pop from `passed` list
  } else {
    continue
  }
}
// Append to passed or failed list
```

Reply:
I thought about checking the limit on the sum, but I think that could cause builds with many passed tests to fill up the clause before we reach the failed test.

So I decided to apply these limits on individual slices instead. The total amount of URI we keep in the accumulator will jump from 1k -> 2k.

Then, in the build event handler, I cramped it back down to ~1k again with a bias to failed test artifacts.

PTAL 🙏

Reply:
Nice, that seems simpler